File: peel/src/main/java/net/tjalp/peach/peel/config/JsonConfig.java
package net.tjalp.peach.peel.config;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;

/**
 * A [JsonConfig] can be used to instantiate
 * parameters in a class without having
 * to manually read the keys and values
 */
public class JsonConfig<T> implements Configurable {

    private final File file;
    private final Class<T> type;
    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();
    private T data;

    public JsonConfig(File file, Class<T> type) {
        this(file, type, true);
    }

    public JsonConfig(File file, Class<T> type, boolean create) {
        this.file = file;
        this.type = type;

        if (file.exists()) {
            load();
        } else {
            try {
                data = type.getConstructor().newInstance();
            } catch (Exception err) {
                throw new IllegalStateException("Failed to instantiate config", err);
            }

            if (create) {
                save(null);
            }
        }
    }

    /**
     * Reload the config from disk and replace the
     * current config.
     */
    private void load() {
        try {
            data = gson.fromJson(Files.newBufferedReader(file.toPath()), type);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Save the current config to disk
     *
     * @param output The output file, null for the input file
     */
    private void save(File output) {
        String json = gson.toJson(data);
        File outFile = output != null ? output : file;

        try {
            outFile.createNewFile();
            Files.write(file.toPath(), json.getBytes());
        } catch (IOException e) {
            e.printStackTrace();
        }

        if (data instanceof Configurable) ((Configurable) data).onSave();
    }

    /**
     * Get the data of the [JsonConfig]
     *
     * @return the class of the JsonConfig
     */
    public T data() {
        return this.data;
    }
}

File: peel/src/main/kotlin/net/tjalp/peach/peel/config/JsonConfig.kt
package net.tjalp.peach.peel.config

import net.tjalp.peach.peel.exception.FailedOperationException
import net.tjalp.peach.peel.util.GsonHelper
import java.io.File

/**
 * Each instance of [JsonConfig] represents
 *
 * Copyright 2019-2021 (c) Exodius Studios. All Rights Reserved.
 *
 * @author Jøøls
 */
class JsonConfig<T : Configurable>(private val file: File, private val type: Class<T>, create: Boolean = true) {

    /** The gson instance **/
    private val gson = GsonHelper.global()

    /** The data contained within the config **/
    lateinit var data: T
        private set

    init {
        if (file.exists()) {
            load()
        } else {
            try {
                data = type.getConstructor().newInstance()
            } catch (err: Exception) {
                throw FailedOperationException("Failed to instantiate config", err)
            }

            if (create) {
                save()
            }
        }
    }

    /**
     * Reload the config from disk and replace the
     * current config.
     */
    fun load() {
        file.bufferedReader().use {
            data = gson.fromJson(it, type)
            data.onLoad()
        }
    }

    /**
     * Save the current config to disk
     *
     * @param output The output file, null for the input file
     */
    fun save(output: File? = null) {
        val json = gson.toJson(data)
        val formatted = GsonHelper.pretty().toJson(json)
        val outFile = output ?: file

        outFile.createNewFile()
        outFile.writeText(formatted)

        data.onSave()
    }

}